core/src/main/java/com/datastax/oss/driver/api/core/type/codec/MappingCodec.java

```diff
@@ -0,0 +1,156 @@
+ *     exist a codec registered for it).
+ * @param <OuterT> The "outer", or target Java type; this codec will handle the mapping to and from
+ *     {@code InnerT} and {@code OuterT}.
+ */
+public abstract class MappingCodec<InnerT, OuterT> implements TypeCodec<OuterT> {
+
+  protected final TypeCodec<InnerT> innerCodec;
+  protected final GenericType<OuterT> outerJavaType;
+
+  /**
+   * Creates a new mapping codec providing support for {@code OuterT} based on an existing codec for
+   * {@code InnerT}.
+   *
+   * @param innerCodec The inner codec to use to handle instances of InnerT; must not be null.
+   * @param outerJavaType The outer Java type; must not be null.
+   */
+  public MappingCodec(
+      @NonNull TypeCodec<InnerT> innerCodec, @NonNull GenericType<OuterT> outerJavaType) {
+    this.innerCodec = Objects.requireNonNull(innerCodec, "innerCodec cannot be null");
+    this.outerJavaType = Objects.requireNonNull(outerJavaType, "outerJavaType cannot be null");
+  }
+
+  /** @return The type of {@code OuterT}. */
+  @NonNull
+  @Override
+  public GenericType<OuterT> getJavaType() {
+    return outerJavaType;
+  }
+
+  /** @return The type of {@code InnerT}. */
+  public GenericType<InnerT> getInnerJavaType() {
+    return innerCodec.getJavaType();
+  }
+
+  /**
+   * Whether this codec is capable of encoding the given Java object.
+   *
+   * <p>This implementation first checks if the value's runtime type is compatible with {@code
+   * OuterT}, then checks if the converted value's runtime type is compatible with {@code InnerT}.
+   *
+   * <p>The object's Java type is inferred from its runtime (raw) type, contrary to {@link
+   * #accepts(GenericType)} which is capable of handling generic types. Therefore, this method can
+   * only handle non-parameterized types; codecs handling parameterized types, such as collection
+   * types, must override this method and perform some sort of "manual" inspection of the actual
+   * type parameters.
+   *
+   * <p>Contrary to other {@code accept} methods, this method's default implementation is
+   * <em>covariant</em> with respect to the passed argument (through the usage of {@link
+   * Class#isAssignableFrom(Class)}) and <em>it's strongly recommended not to modify this
+   * behavior</em>. This means that, by default, a codec will accept <em>any subtype</em> of the
+   * Java type that it has been created for. This is so because codec lookups by arbitrary Java
+   * objects only make sense when attempting to encode, never when attempting to decode, and indeed
+   * the {@linkplain #encode(Object, ProtocolVersion) encode} method is covariant with {@code
+   * JavaTypeT}.
+   */
+  @Override
+  public boolean accepts(@NonNull Object value) {
+    Preconditions.checkNotNull(value);
+    if (getJavaType().getRawType().isAssignableFrom(value.getClass())) {
+      try {
+        @SuppressWarnings("unchecked")
+        OuterT outer = (OuterT) value;
```

Comment:
if someone will create a custom codec and mess with types he will get `CastException` right? Maybe we could treat that `Exception` differently and log it?

Reply:
The cast here cannot fail, since we have a check for `getJavaType().getRawType().isAssignableFrom(value.getClass())` earlier.